text/0141-linking-traces.md

```diff
@@ -0,0 +1,368 @@
+
+## Additional Use Cases
+
+In addition, besides inter-trace links, we might also be interested in linking spans within one trace, for example to link spans that are related to each other but not in a direct (parent-child or linear) hierarchy. While this is not the main objective of the RFC, our proposed solution would also facilitate this linkage. Some concrete examples (but out of scope):
+- (Async) data flow operations (e.g. RxJS pipes) where we would be able to link individual operations.
+- Websocket spans where we could link individual web socket messages
+- Background data syncs on mobile that you can't directly link to a specific transaction, but it's useful to know that they happened during a transaction. Now, the mobile SDKs put these spans on the transaction bound to the scope.
+- More generally, whenever we want to "casually" establish a relationship between spans that cannot or should not be achieved via a hard relationship (traceId or parent spanId)
+
+# Concrete Goals of achieving linked traces
+
+Once this RFC is accepted, we have settled on all important details needed to implement connected traces.
+This means, we have all the necessary information to support the following concrete goals:
+
+TODO: List concrete product/UI goals
+
+- SDKs can add support for adding links to spans
+- We can ingest and store span links
+- Frontend SDKs can automatically add span links so that the current trace root span links back to the previous trace root span
+
+
+## Secondary goals
+
+- We can shorten the lifetime of frontend traces because we're now able to link to the previous trace instead. This is something we'll do but it's a follow-up and can only be addressed once the primary goals are achieved. 
+
+## Out of scope/Non-Goals
+
+- Querying across linked traces: Given the "linked list" nature of one root span linking back to its previous root span, we accept that obtaining a complete list is an expensive query operation. Therefore, we consider use cases that require a complete list of all linked traces out of scope. We can address this as a follow-up if technically possible but we're aware that this is likely a sub-optimal data structure for such queries. 
+- Linking transactions started during the rendering of statically generated HTML pages and pageload transactions. A span link is ideal here because we don't want the hard link of a trace id. However, a casual link would still increase context and completeness. We might do it but not at this point as it's a follow-up item.
+- Unsampled traces are not linked. Errors across unsampled traces are not linked. For example, if the first trace was sampled negatively, the next trace will not link to it as this would result in a link to a non-existing trace.
```

Comment:
The SDK can manage that for client sampling, but how does dynamic sampling affect this? 🤔

Reply:
This is a good point. I think similarly to parent span id's we cannot assume that just because a link to a span exists, said span also exists. Meaning, for a simple previous/next button scenario in the trace view, we'd actually need to search for the trace with the linked span before displaying the button(s). This ties in well with the EPD decision that the product has to handle incomplete traces. Or IOW, SDKs not linking to already SDK-dropped spans is not enough but should reduce the number of non-existing links.

Reply:
I've removed the last point in this section, since we decided in our last call that the SDK will send a span link to a previously negatively sampled root span to indicate that there was a negatively sampled trace (see the diagram I added below).